src/lib/krb5/os/dnssrv.c

```diff
@@ -45,6 +45,122 @@ krb5int_free_srv_dns_data (struct srv_dns_entry *p)
+    k5_buf_add_len(&buf, realm->data, realm->length);
+
+    /* Realm names don't (normally) end with ".", but if the query
+       doesn't end with "." and doesn't get an answer as is, the
+       resolv code will try appending the local domain.  Since the
+       realm names are absolutes, let's stop that.
+
+       But only if a name has been specified.  If we are performing
+       a search on the prefix alone then the intention is to allow
+       the local domain or domain search lists to be expanded.  */
+
+    if (buf.len > 0 && fixed_buf[buf.len - 1] != '.')
+        k5_buf_add(&buf, ".");
+
+    return k5_buf_status(&buf);
+}
+
+/* Insert into the list, ordering by priority. Weight is not currently
+ * used. */
+static void
+place_srv_entry(struct srv_dns_entry **head, struct srv_dns_entry *new)
+{
+    struct srv_dns_entry *entry;
+
+    if (*head == NULL || (*head)->priority > new->priority) {
+        new->next = *head;
+        *head = new;
+        return;
+    }
+    /* Insert an entry into this spot if the next entry has a higher
+     * priority (lower priorities are preferred), or if there is no
+     * next entry (we're at the end).  */
+    for (entry = *head; entry != NULL; entry = entry->next) {
+        if ((entry->next &&
+             entry->next->priority > new->priority) ||
+            entry->next == NULL) {
+            new->next = entry->next;
+            entry->next = new;
+            break;
+        }
+    }
+}
+
+/* Query the URI RR, collecting weight, priority, and target. */
+krb5_error_code
+k5_make_uri_query(const krb5_data *realm, const char *service,
+                  struct srv_dns_entry **answers)
+{
+    const unsigned char *p = NULL, *base = NULL;
+    char host[MAXDNAME];
+    int size, ret, rdlen;
+    unsigned short priority, weight;
+    struct krb5int_dns_state *ds = NULL;
+    struct srv_dns_entry *head = NULL, *uri = NULL;
+
+    *answers = NULL;
+     /* Construct service.realm. */
+    ret = prepare_lookup_buf(realm, service, NULL, host, sizeof(host));
+    if (ret)
+        return 0;
+
+#ifdef TEST
+    fprintf (stderr, "sending DNS URI query for %s\n", host);
+#endif
+
+    size = krb5int_dns_init(&ds, host, C_IN, T_URI);
+    if (size < 0)
+        goto out;
+
+    for (;;) {
+        ret = krb5int_dns_nextans(ds, &base, &rdlen);
+        if (ret < 0 || base == NULL)
+            goto out;
+
+        p = base;
+
+        SAFE_GETUINT16(base, rdlen, p, 2, priority, out);
+        SAFE_GETUINT16(base, rdlen, p, 2, weight, out);
+
+        uri = k5alloc(sizeof(*uri), &ret);
+        if (uri == NULL)
+            goto out;
+
+        uri->priority = priority;
+        uri->weight = weight;
+        /* Do rdlen - 4 due to the priority and weight. */
+        uri->host = k5memdup0(p, rdlen - 4, &ret);
+        if (uri->host == NULL) {
+            ret = errno;
+            goto out;
+        }
+
+        place_srv_entry(&head, uri);
+    }
+
+out:
+    if (ds != NULL) {
```

Comment:
krb5int_dns_fini() handles null pointers, so don't conditionalize the call.  Also don't assign ds = NULL right before the end of the function.  (k5_try_realm_txt_rr() has the same issue.  But when adding new code which mirrors existing code, don't mirror the bad or outdated practices in the old code.)